src/ShopifyApp/Middleware/AuthShop.php

```diff
@@ -52,6 +52,12 @@ protected function validateShop(Request $request)
         $shopDomainSession = $session->getDomain();
         $shopDomain = ShopifyApp::sanitizeShopDomain($shopDomainParam ?? $shopDomainSession);
 
+        // See issue https://github.com/ohmybrew/laravel-shopify/issues/295
+        parse_str(parse_url($request->header('referer'), PHP_URL_QUERY), $refererQueryParams);
+        if (isset($refererQueryParams['shop']) && $shopDomain !== $refererQueryParams['shop'] && ShopifyApp::api()->verifyRequest($refererQueryParams)) {
```

Comment:
@ohmybrew if I don't verify the params then I can modify the referer domain to be the other shops domain and it loads the modified domain (basically make the problem exhibit itself again) and makes it possible for an attacker to gain control of a shop through IDOR.

I will adjust the code to make it only verify if it really and truly needs to and add the tests when they're done.

OWASP A4 - [IDOR - Insecure Direct Object References](https://www.owasp.org/index.php/Top_10_2013-A4-Insecure_Direct_Object_References)

There would have been concerns about performance if verifying the request through the API called the server, but I checked the code in `OhMyBrew/BasicShopifyAPI.php` and it just does a signature check - it doesn't make any network requests.

It's important to not trust data received from the client, and to sanitise and check actions like this.